coil-gif/src/main/java/coil3/decode/AnimatedImageDecoderDecoder.kt

```diff
@@ -195,5 +195,14 @@ class AnimatedImageDecoderDecoder @JvmOverloads constructor(
     }
 }
 
+internal fun BufferedSource.squashToDirectByteBuffer(): ByteBuffer {
+    // TODO: How can we know BufferedSource's length, avoid allocating that bytearray?
+    val bytes = readByteArray()
```

Comment:
@colinrtwhite 
Is it acceptable to add a length field for [ImageSource] and let fetcher initialize it
To avoid allocating this ByteArray, we use https://docs.oracle.com/javase/8/docs/api/java/nio/channels/ReadableByteChannel.html#read-java.nio.ByteBuffer-
write to DirectByteBuffer directly?

Reply:
We can't add a `length` field to `ImageSource` as it's not possible to support consistently. Adding it would mean supporting it inconsistently across our fetchers. I'd also want to avoid adding `length` just to support this use case.

Unfortunately this might prevent us from adding this optimization (since we have to allocate the byte array anyways).

Reply:
No matter, there is another way to achieve this